Allow runtimes to be added and default runtime to be modified on config reload

elezar opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds support for reloading runtime configs on SIGHUP. This allows for runtimes to be added dynamically and the default runtimes to be adjusted.

This is useful for cases where runtimes are added / configured after crio is started. One example of this is the NVIDIA Container Runtime which is managed by the NVIDIA GPU Operator. Without this change, adding a runtime to the config requires a node restart which is disruptive.

#### Which issue(s) this PR fixes:

Fixes #6036

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Allow for a reload to add additional runtimes and change the default runtime
```

TomSweeneyRedHat commented:
The code LGTM for what you're intending to do.  However, I'll have to lean on @giuseppe , @haircommander and @saschagrunert on the safeness of swapping out the runtime like this.

haircommander commented on pkg/config/reload.go:
```diff
@@ -228,3 +231,32 @@ func (c *Config) ReloadRdtConfig(newConfig *Config) error {
+		if _, ok := c.Runtimes[runtime]; !ok {
+			continue
+		}
+		newRuntimes = append(newRuntimes, runtime)
+	}
+
+	for _, runtime := range newRuntimes {
+		c.Runtimes[runtime] = newConfig.Runtimes[runtime]
+	}
```
couldn't we just compress these lines and do 
`		c.Runtimes[runtime] = newConfig.Runtimes[runtime]` if it wasn't found?

haircommander commented on pkg/config/reload.go:
```diff
@@ -228,3 +231,32 @@ func (c *Config) ReloadRdtConfig(newConfig *Config) error {
+	var newRuntimes []string
+	for runtime := range newConfig.Runtimes {
+		if _, ok := c.Runtimes[runtime]; !ok {
+			continue
```
I think we should log a warning that we skipped because it was already present

haircommander commented on pkg/config/reload_test.go:
```diff
@@ -325,4 +326,40 @@ var _ = t.Describe("Config", func() {
+			Expect(err).NotTo(BeNil())
+		})
+
+		It("should add a new runtime", func() {
```
could we add a case where we verify a runtime is unchanged because it's already present?

elezar replied:
Sure. I will add more test cases. I wanted to get this out for early feedback.

haircommander commented on pkg/config/config.go:
```diff
@@ -1089,6 +1076,27 @@ func (c *RuntimeConfig) Validate(systemContext *types.SystemContext, onExecution
+	// check we do have at least a runtime
+	if _, ok := c.Runtimes[c.DefaultRuntime]; !ok {
+		// Set the default runtime to "runc" if default_runtime is not set
+		if c.DefaultRuntime == "" {
```
nit, can we invert this condition and compress the else?

`if c.DefaultRuntime != "" { return } ...`

haircommander replied:
I suppose this was already there, but doesn't hurt to update now 🙂

saschagrunert reviewed (commented):
The unit test failures may be legit. Can we also add a test to https://github.com/cri-o/cri-o/blob/main/test/reload_config.bats, maybe by injecting a fake runtime?

elezar commented:
> The unit test failures may be legit. Can we also add a test to https://github.com/cri-o/cri-o/blob/main/test/reload_config.bats, maybe by injecting a fake runtime?

OK. I will have a look.

elezar commented:
@saschagrunert I have updated the unit tests and also added some tests as suggested. I don't have my development environment quite set up, so I've been resting on the CI a bit more heavily than I probably should. Please take another look (especially at the tests) and let me know if I'm missing something.

I'll try to get my dev env set up a bit more cleanly if more changes are required.

haircommander commented:
would you still consider this to be a draft @elezar ?

elezar commented:
> would you still consider this to be a draft @elezar ?

@haircommander no, I don't think so. I just wanted to get the tests looking in a better shape. Not having a local dev environment set up is making that a bit slow.

elezar commented:
I have rebased the changes here and updated the status.

What are the best docs for getting a development environment set up? Ideally this would be containerized.

elezar commented on pkg/config/reload.go:
```diff
@@ -237,3 +240,36 @@ func (c *Config) ReloadRdtConfig(newConfig *Config) error {
+
+	if c.DefaultRuntime != newConfig.DefaultRuntime {
+		c.DefaultRuntime = newConfig.DefaultRuntime
+		if err := c.ValidateDefault(); err != nil {
```
This was a typo after a rebase. Should be fixed now.

klihub commented:
/test kata-containers

elezar commented:
/test integration_rhel

haircommander commented:
@elezar is this ready?

/retest

haircommander commented:
/approve

LGTM, thanks @elezar !

@cri-o/cri-o-maintainers PTAL

zvonkok commented:
@mrunalp PTAL, if this gets merged, in which OCP version can we expect this to land?

haircommander commented:
> @mrunalp PTAL, if this gets merged, in which OCP version can we expect this to land?

main is currently feeding into 1.26, so 4.13 would be the target

@saschagrunert PTAL

saschagrunert commented on pkg/config/reload.go:
```diff
@@ -237,3 +240,36 @@ func (c *Config) ReloadRdtConfig(newConfig *Config) error {
+	}
+
+	if err := c.ValidateRuntimes(); err != nil {
+		return fmt.Errorf("unabled to reload runtimes: %v", err)
```
```suggestion
		return fmt.Errorf("unabled to reload runtimes: %w", err)
```

elezar replied:
Updated and rebased.

saschagrunert reviewed (commented):
Just one tiny error wrapping nit, otherwise LGTM

elezar commented:
/retest-required

haircommander commented:
thanks for your patience as we went through a CI migration and multiple outages

/lgtm

haircommander commented:
oh bother
/hold

elezar commented:
/retest

haircommander commented:
/hold cancel

elezar commented:
/retest

elezar commented:
@haircommander thanks for giving me the heads-up regarding the CI issues. I have rebased the changes off `main` again. Are CI issues still expected?

haircommander commented:
/retest
